Grid view optimization

dstandish opened the pull request:
Work in progress branch for grid optimization

I break up the monolith that is grid_data.

The headline here is, with 6k tasks in a dag, loading time for 10 runs drops from 1.5m to < 10s in a quick local test.

I split it into smaller more purpose-specific requests that each do _less_.  So we have one request for just the structure, and another one for TI states (per dag run).  I also find ways to stop refreshing when there's no active dag run (or the particuler dag run is not active and _its_ tis don't need refreshing.  I also changed the "latest dag run" query (which checks for a new run triggered externally to be simpler dedicated endpoint.  It runs  ever couple seconds even when there is nothing going on and now it takes 10ms instead of 300ms.

In order to have the grid/structure request stop refreshing when nothing is active i had to add a new context provider so the state could be propagated from the grid/runs request (which knows if there's active runs).   There may be a better way to do this, and the linter may not like it.

Here's an dag you can test this with.  Before, if you had 10 runs, it would take 1.5 minutes to load.  When I tried, it was 6 seconds.  It's faster with non-dev mode than with dev mode presumably because of parallelism.


```python
from __future__ import annotations

from airflow.providers.standard.operators.empty import EmptyOperator
from airflow.sdk import DAG, TaskGroup, chain

with DAG("bighello_deeper_only_normal"):
    for i in range(10):
        with TaskGroup(f"group_{i}"):
            EmptyOperator(task_id="hello")
            with TaskGroup(f"group_{i}2"):
                chain([EmptyOperator(task_id=f"empty_{j}") for j in range(100)])
                chain([EmptyOperator(task_id=f"empty2_{j}") for j in range(100)])
                # EmptyOperator.partial(task_id=f"hello2").expand(doc=list(range(100)))
            with TaskGroup(f"group_{i}3"):
                chain([EmptyOperator(task_id=f"empty_{j}") for j in range(100)])
                chain([EmptyOperator(task_id=f"empty2_{j}") for j in range(100)])
                # EmptyOperator.partial(task_id=f"hello2").expand(doc=list(range(100)))
            with TaskGroup(f"group_{i}4"):
                chain([EmptyOperator(task_id=f"empty_{j}") for j in range(100)])
                chain([EmptyOperator(task_id=f"empty2_{j}") for j in range(100)])
                # EmptyOperator.partial(task_id=f"hello2").expand(doc=list(range(100)))
```

## before (1.6m)

<img width="1269" alt="before-grid-optimization" src="https://github.com/user-attachments/assets/a8968d6f-651e-4d69-ab72-a5daf7387c6e" />

## after (6s)

<img width="1273" alt="after-grid-optimization" src="https://github.com/user-attachments/assets/a29bb979-2b51-46b9-a8eb-95c34ded833e" />

## in action
https://github.com/user-attachments/assets/54e42f62-57a2-46e4-87f2-262419a16ab9

jedcunningham commented on airflow-core/src/airflow/api_fastapi/core_api/datamodels/ui/common.py:
docstrings need a little love here

jedcunningham commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py:
```diff
@@ -257,8 +268,254 @@ def grid_data(
+
+    # Retrieve, sort the previous DAG Runs
+    base_query = select(DagRun.id).where(DagRun.dag_id == latest_dag.dag_id)
+    # This comparison is to falls to DAG timetable when no order_by is provided
```
```suggestion
    # This comparison is to fall back to DAG timetable when no order_by is provided
```

jedcunningham commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py:
```diff
@@ -257,8 +268,254 @@ def grid_data(
+        DagRun.run_type,
+    ).where(DagRun.dag_id == latest_dag.dag_id)
+
+    # This comparison is to falls to DAG timetable when no order_by is provided
```
```suggestion
    # This comparison is to fall back to DAG timetable when no order_by is provided
```

jedcunningham commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py:
```diff
@@ -257,8 +268,254 @@ def grid_data(
+    ).where(DagRun.dag_id == latest_dag.dag_id)
+
+    # This comparison is to falls to DAG timetable when no order_by is provided
+    if order_by.value == order_by.get_primary_key_string():
```
Does it make sense to defer looking up the latest dag serdag at all, unless we need the default order_by? Probably worth a follow up at this point though.

dstandish replied:
good eye

jedcunningham commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py:
```diff
@@ -257,8 +268,254 @@ def grid_data(
+        limit=None,
+        return_total_entries=False,
+    )
+    # todo: so, what we need to do is
```
Is this a todo or just informational?

jedcunningham commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py:
```diff
@@ -257,8 +268,254 @@ def grid_data(
+    dag_id: str,
+    session: SessionDep,
+) -> LatestRunResponse | None:
+    """Return unified dag structure for grid view."""
```
```suggestion
    """Return the latest run of a dag."""
```

jedcunningham commented on airflow-core/src/airflow/api_fastapi/core_api/services/ui/grid.py:
```diff
@@ -309,3 +311,204 @@ def _get_node_by_id(nodes, node_id):
+    """
+    # Additional logic to calculate the overall states to cascade recursive task states
+    # todo: must be a faster way
+    #   it also doesn't necessarily make sense if the last state in the group is success
```
I wonder if we can just look at the leaf states? Because yeah, looking at everything isn't great - something in the middle could "fail", but be handled by a downstream via trigger rule...

jedcunningham commented on airflow-core/tests/unit/api_fastapi/core_api/routes/ui/test_grid.py:
```diff
@@ -1408,3 +1437,79 @@ def test_should_response_200_with_deleted_task_and_taskgroup(self, session, test
+        #         {"task_id": "task_group.inner_task_group.inner_task_group_sub_task"},
+        #         {"task_id": "task_group.mapped_task"},
+        #     ],
+        # }
```
```suggestion
```

dstandish committed:
Refactor grid to use multiple endpoints

The headline here is, with 3k tasks in a dag, loading time for 10 runs drops from 1.5m to < 10s in a quick local test.

I split it into smaller more purpose-specific requests that each do less. So we have one request for just the structure, and another one for TI states (per dag run). I also find ways to stop refreshing when there's no active dag run (or the particuler dag run is not active and its tis don't need refreshing. I also changed the "latest dag run" query (which checks for a new run triggered externally to be simpler dedicated endpoint. It runs ever couple seconds even when there is nothing going on and now it takes 10ms instead of 300ms.

In order to have the grid/structure request stop refreshing when nothing is active I had to add a new context provider so the state could be propagated from the grid/runs request (which knows if there's active runs). There may be a better way to do this, and the linter may not like it.

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py:
```diff
@@ -257,8 +268,254 @@ def grid_data(
+    ],
+    response_model_exclude_none=True,
+)
+def get_dag_structure(
```
There is an endpoint to achieve exactly that I believe `def structure_data`.


With `external_dependencies=False` it should be similar. Also it handles versions.

dstandish replied:
it is used for a different purpose and does different things.  my approach is to be purpose-specific and minimal, and that's why it is fast.

dstandish replied:
more directly... that other endpoint is certainly doing "more" than this one and, i don't think it makes sense to combine just for the sake of having fewer endpoints -- this one is purpose-built and does the job and i don't think it's worth the trade off to couple it to the other use case.

dstandish replied:
@bbovenzi i agree and this is already the case -- this endpoint is actually `/grid/structure`. it's just not obvious because it's using `@grid_router`

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/services/ui/grid.py:
```diff
@@ -309,3 +311,204 @@ def _get_node_by_id(nodes, node_id):
     return {}
+
+
+def fill_task_instance_summaries_simple(
```
What do we imply by `simple` (I guess it's in regards to the current `fill_task_instance_summaries`)? It does not seem used at the moment.

dstandish replied:
it might be an earlier attempt  -- i have to do some cleanup and polishing

dstandish replied:
removed

pierrejeambrun commented on airflow-core/src/airflow/utils/task_group.py:
```diff
@@ -94,3 +94,44 @@ def task_group_to_dict(task_item_or_group, parent_group_is_mapped=False):
     }
+
+
+def task_group_to_dict_grid(task_item_or_group, parent_group_is_mapped=False):
```
Maybe we can find a way to reuse `task_group_to_dict` seems pretty similar.

dstandish replied:
they are similar but the other one does something different re children.... you could DRY it up just a bit but it would add a little complexity.  Personally i think it's better to avoid the coupling and just let the logic be independent for the diff use case.  It's just an internal helper function that retrieves the info needed for the _grid_ use case as opposed to the _graph_ use case.

pierrejeambrun reviewed (commented):
I love the general approach by splitting things into smaller request to not block UI rendering. Results are impressive!



It needs some polishing but I think we are going in the right direction.

That's a big effort, thanks Daniel for taking this one.

bbovenzi commented on airflow-core/src/airflow/ui/src/pages/DagRuns.tsx:
```diff
@@ -167,7 +167,7 @@ export const DagRuns = () => {
       dagId: dagId ?? "~",
       endDateLte: endDate ?? undefined,
-      limit: pagination.pageSize,
+      limit: 123,
```
Wait what?

dstandish replied:
haha sorry that was when i was trying to debug which call was being made from where 🤪

bbovenzi commented on airflow-core/src/airflow/ui/src/queries/useGridStructure.ts:
```diff
@@ -0,0 +1,47 @@
+import { HasActiveRunContext } from "src/layouts/Details/context.ts";
+import { useAutoRefresh } from "src/utils";
+
+export const useGridStructure = (limit: number) => {
```
Nit, let's make this an object so we can add more params later and also know what `(1)` is doing.

```suggestion
export const useGridStructure = ({ limit}: { limit: number }) => {
```

dstandish replied:
good suggestion. done.

bbovenzi commented on airflow-core/src/airflow/ui/src/queries/useGridStructure.ts:
```diff
@@ -0,0 +1,47 @@
+      dagId,
+      limit,
+      orderBy: "-run_after",
+      // runAfterLte: runAfter, // todo: this doesn't make sense if sorting by logical date
```
Todo to add it back or remove it?

dstandish replied:
this comment is from earlier as i was working my way through the code. right now we aren't filtering by runAfter, and i'm not really sure we were before either

dstandish replied:
i'll remove the comment.

dstandish committed:
replace grid_data with ti_summaries in the group details

dstandish committed:
Revert "replace grid_data with ti_summaries in the group details"

This reverts commit e6f09a88aae9ec42fca70b24009fe6bac2b4da8b.

dstandish committed:
formatting

dstandish committed:
restore the button to refresh if selected run not visible

dstandish committed:
Remove the context provider for refresh signaling

dstandish committed:
pre-commit fixes

dstandish committed:
fix test

dstandish committed:
remove commented code

dstandish committed:
docstrings etc

ashb commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py:
```diff
@@ -257,8 +268,324 @@ def grid_data(
+
+
+@grid_router.get(
+    "/structure/{dag_id}",
```
Nit it is not too late (but also don't let this stop merging!)

```suggestion
    "/{dag_id}/structure",
```

As the structure is "a property" of the dag this feels more natural to me

dstandish replied:
i'll do these as followup

ashb commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py:
```diff
@@ -257,8 +268,324 @@ def grid_data(
+
+
+@grid_router.get(
+    "/runs/{dag_id}",
```
Ditto etc

dstandish replied:
i'll do these as followup

dstandish committed:
stop refreshing structure if no pending run

dstandish committed:
fix typing

dstandish committed:
fix static checks
